### PR TITLE
net: don't bind on 0.0.0.0 if binds are restricted to Tor

### PR DESCRIPTION
**Pull request by vasild**
The semantic of `-bind` is to restrict the binding only to some address.
If not specified, then the user does not care and we bind to `0.0.0.0`.
If specified then we should honor the restriction and bind only to the 
specified address.

Before this change, if no `-bind` is given then we would bind to
`0.0.0.0:8333` and to `127.0.0.1:8334` (incoming Tor) which is ok -
the user does not care to restrict the binding.

However, if only `-bind=addr:port=onion` is given (without ordinary
`-bind=`) then we would bind to `addr:port` _and_ to `0.0.0.0:8333` in
addition.

Change the above to not do the additional bind: if only
`-bind=addr:port=onion` is given (without ordinary `-bind=`) then bind
to `addr:port` (only) and consider incoming connections to that as Tor 
and do not advertise it. I.e. a Tor-only node.

### CODE REVIEW

**Comment by vasild** [comment]
This is a followup to #19991, cc @hebasto

**Comment by luke-jr** [comment]
How does Tor continue to function after this?

**Comment by JabbaDesilijicTiure** [comment]
I don't think you need to bind Tor to anywhere. Tor in fact is very dangerous if bridges are not used, without bridges Tor allows a government or any other intelligence serves to see the Tor traffic which in some cases result in imprisonment or death. 

Tor is already bound to 127.0.0.1:**9050**. You could just tell the user to edit the torrc file and allow Tor control at 127.0.0.1:**9051**

Obviously, I don't get it why would bitcoin need to bind Tor to anywhere, and especially I don't understand why this PR is marked with the P2P tag. Neither Tor not Bitcoin is P2P, both Bitcoin and Tor are decentralized but NOT P2P. Some people don't understand what is peer-to-peer far as I see. 

Communication between decentralized servers are communication like any other between any other servers, in no way this is peer-to-peer. A real peer-to-peer would allow to send and receive data without any central server or network by any member of the network and this must be the case even **offline**.

There are 3 types of networks, centralized like bitcoin.org, decentralized like the bitcoin software, and distributed. Only distributed network can be called as P2P. **If Bitcoin would be peer-to-peer, than mining of Bitcoin would be distributed too** and each user can independently from any pool would able to mine Bitcoin with the help of all other nodes! 

The only P2P in Bitcoin is the send and receive function anything else is either decentralized or centralized. 

My advise is to keep Tor out of Bitcoin core.

**Comment by jonatack** [comment]
Concept ACK

File: test/functional/feature_bind_extra.py
Comment (jonatack) [inline test/functional/feature_bind_extra.py:114]: This assertion fails for me with errors like
```
AssertionError: not(
{('00000000', 11340), ('7f000001', 16340), ('00000000000000000000000000000000', 11340), ('7f000001', 18445)} ==
{('00000000', 11340), ('7f000001', 16340), ('7f000001', 18445)})
```
This fixes it, but it's not pretty and might be treating the symptoms more than the problem:
```diff
--- a/test/functional/test_framework/netutil.py
+++ b/test/functional/test_framework/netutil.py
@@ -81,7 +81,10 @@ def get_bind_addrs(pid):
     bind_addrs = []
     for conn in netstat('tcp') + netstat('tcp6'):
         if conn[3] == STATE_LISTEN and conn[4] in inodes:
-            bind_addrs.append(conn[1])
+            if conn[1][0] == '00000000000000000000000000000000':
+                bind_addrs.append(('00000000', conn[1][1]))
+            else:
+                bind_addrs.append(conn[1])
     return bind_addrs
```

File: test/functional/feature_bind_extra.py
Comment (jonatack) [inline test/functional/feature_bind_extra.py:2]:
```suggestion
# Copyright (c) 2020 The Bitcoin Core developers
```

**Review by jonatack — COMMENTED** [review]
Approach ACK/almost ACK db2d6f323836f286f37d4e826c6edcfce6cde29b

File: test/functional/feature_bind_extra.py
Comment (JabbaDesilijicTiure) [reply test/functional/feature_bind_extra.py:2]: @jonasnick actually @vasild has all the copyright to this file so it should be: 

`Copyright (c) 2020 Vasil Dimov`

Just as Satoshi Nakamoto has all copyright in this software from 2008-2010! 
Your recommendations on the pull request doesn't make the file the property of "The Bitcoin Core developers". 
You need to change significant part of the present file to claim that the "Bitcoin Core developers" own the copyright, until then Vasil Dimov has all copyright.

Comment (JabbaDesilijicTiure) [inline test/functional/feature_bind_extra.py:2]:
```suggestion
# Copyright (c) 2020 Vasil Dimov
```

Comment (JabbaDesilijicTiure) [reply test/functional/feature_bind_extra.py:2]: Example: 

https://github.com/bitcoin/bitcoin/blob/db2d6f323836f286f37d4e826c6edcfce6cde29b/src/secp256k1/src/ecdsa.h#L1-L5

Above you can see that @sipa also retained all his copyright and his pull request and merge doesn't automatically made his work the property of the "Bitcoin Core developers".

Comment (sipa) [reply test/functional/feature_bind_extra.py:2]: I am not a lawyer, but my understanding is that:
* The copyright notice here has no legal meaning, and copyright is retained by whoever authored it regardless of what is stated in the file's header.
* "Bitcoin Core developers" is no legal entity, and can't own any copyright. It's just there to indicate "various contributors over time" without the need for being specific.

So there is no problem with using either your own name, or using the collective term - except the latter avoids the need for keeping it up to date as more contributors work on it. Most code I've written for Bitcoin Core is in files with "Bitcoin Core developers" notices, FWIW.

Comment (JabbaDesilijicTiure) [reply test/functional/feature_bind_extra.py:2]: > The copyright notice here has no legal meaning, and copyright is retained by whoever authored it regardless of what is stated in the file's header.

The copyright notice is required. Satoshi Nakamoto stated that his code is released under MIT which requires you to keep his name and his orginal copyright notice in place, so when you make changes in the same files created by Satoshi Nakamoto you are required to add your name like this: 

```
Copyright (c) 2008-2010 Satoshi Nakamoto
Copyright (c) 2011-2013 Pieter Wuille
and here comes the MIT statement 
```

The copyright years can't overlap, like you can't say Copyright (c) 2008-2013 Pieter Wuille when there is the clear indication of the years beside the name of Satoshi Nakamoto, this would be a serious legal problem for you. If you don't respect other peoples copyright nobody will respect yours, why would they. 

Saying that copyright notice has no legal meaning is absurd, and under some circumstances if your words turn into action even considered as criminal under international law.

Comment (luke-jr) [reply test/functional/feature_bind_extra.py:2]: This PR is by @vasild. He can legally put any copyright line he wants.

And again, there is no "Bitcoin Core developers" entity. It simply means each author retains his own copyright.

Comment (JabbaDesilijicTiure) [reply test/functional/feature_bind_extra.py:2]: @luke-jr if there is no such entity as "Bitcoin Core developers", than this software has very serious copyright issues and each and all contributor by name must be listed in all copyright notices. If there is no _such entity as "Bitcoin Core developers"_ and as @sipa said _copyright notice here has no legal meaning_ than the entire Bitcoin core software is under the **public domain**. 

The MIT license by Satoshi Nakamoto doesn't matter at all because it has "no legal meaning". 

On this end I would strongly suggest to simply just remove all meaningless copyright notices which assigned to an non-existent entity and persons.

Comment (sipa) [reply test/functional/feature_bind_extra.py:2]: @JabbaDesilijicTiure My comment was purely about the individual files' copyright notices. The project falls under the MIT license as specified in the COPYING file at the root of the repository.

Comment (luke-jr) [reply test/functional/feature_bind_extra.py:2]: @JabbaDesilijicTiure You don't know what you're talking about. Copyright does not depend on any notice at all. There are plenty of places you can learn more - this isn't the place for it.

Comment (JabbaDesilijicTiure) [reply test/functional/feature_bind_extra.py:2]: luke-jr-the-toothless-bitcoin-core-dev-copyright

lol

You didn't learned on IRC that I'm not a fan of you? I'm that guy who owns the **https://bitcoin.foundation** domain and I told you that I will make your life very difficult. 

I'm at least **toothsome**! If you don't know what that mean take a mirror front of you or look it up in a dictionary.

Comment (fanquake) [reply test/functional/feature_bind_extra.py:2]: @JabbaDesilijicTiure I've removed your comment. This repository is obviously not the place for that.

**Comment by laanwj** [comment]
@vasild, like any contributor, is **part of** "The Bitcoin Core developers"
This is not the place for copyright discussion in any case. Please keep the discussion focused on the code change done here.

**Comment by gmaxwell** [comment]
Unless I am misunderstanding this PR, this will break tor configuration for any user with an explicit bind configuration.  Not binding to localhost is extremely weird, I can't think of any other program I've encountered doing that.

Is there an actual usecase for not binding localhost?

Might it be better to just document that localhost binds are implicit and then, if there is a need for not binding to localhost, have an argument to specifically disable that?

In other words, I think what you're actually pointing to is just a doc bug where the documentation isn't pedantically correct, but where the behaviour is what 99.999% of users would prefer. If so-- change the doc.

**Comment by JabbaDesilijicTiure** [comment]
> @vasild, like any contributor, is **part of** "The Bitcoin Core developers"
> This is not the place for copyright discussion in any case. Please keep the discussion focused on the code change done here.

@luke-jr said that there is no such entity as "Bitcoin Core developers" and I clearly noted that the copyright notice should bear the name of @vasild because he is the sole author, not a fictional entity.

To keep my comments on the topic @gmaxwell pointed out that "**not binding to localhost is extremely weird**". That is correct. 
I actually fail to understand why bitcoind can't have a connection to the Tor control port on **127.0.0.1:9051**, that way users who wish to run bitcoind in "Tor-only" mode can get a new .onion address each times the service starts or restarts for outgoing connections. Or to have a fixed onion address in incoming connections.

Basically, you binding to 127.0.0.0:whateverport (maybe even random port) and than using Tor controller to open for incoming Tor connections on the bitcoincorenode0**.onion**:8333, for outgoing obviously you don't need to open any port while in Tor only mode.

**Comment by vasild** [comment]
@gmaxwell 

> this will break tor configuration for any user with an explicit bind configuration. Not binding to localhost is extremely weird

You found a bug! Not in this PR, but in 0.20.1:

```
$ bitcoind --version
Bitcoin Core version v0.20.1.0-36355f128f8b

$ bitcoind -bind=1.1.1.1:8333 -listenonion=1
...
2020-10-29T11:51:58Z tor: Got service ID 7hgc4pgtuexasesn, advertising service 7hgc4pgtuexasesn.onion:8333
...
```

`bitcoind` tells the Tor daemon to redirect incoming connections to `127.0.0.1:8333`:

https://github.com/bitcoin/bitcoin/blob/v0.20.1/src/torcontrol.cpp#L539

but then it does not listen on `127.0.0.1:8333`:

```
$ sockstat -l |grep bitcoin
vd       bitcoind   95898 6  tcp4   127.0.0.1:8332        *:*
vd       bitcoind   95898 21 tcp4   1.1.1.1:8333          *:*
$ telnet 127.0.0.1 8333
Trying 127.0.0.1...
telnet: connect to address 127.0.0.1: Connection refused
```

That bug was fixed by #19991 by removing the hardcoded `127.0.0.1`.

When only `-bind=` is provided (without `-bind=...=onion`), this PR would keep the same behavior as 0.20.1 as above - it would bind only on `1.1.1.1:8333` as requested. However it would tell the Tor daemon to redirect connections to `1.1.1.1:8333` instead of the bogus `127.0.0.1:8333` - as a side effect of the change from #19991 now we use the proper address instead of hardcoding `127.0.0.1`.

**Comment by gmaxwell** [comment]
rpc bind is a separate configuration from 'bind', as is whitebind, perhaps the onion should be its own bind statement (that users normally never touch)

**Comment by JabbaDesilijicTiure** [comment]
@vasild **1.1.1.1:8333 should be not used to bind bitcoind!** 1.1.1.1 is an IP address of CloudFlare's DNS resolver! Use instead **127.0.0.2**:8333

**Comment by JabbaDesilijicTiure** [comment]
@gmaxwell as you see @vasild binding bitcoind to CloudFlare DNS resolver instead to 127.0.0.x. I would recommend not to merge this pull request.

File: test/functional/feature_bind_extra.py
Comment (vasild) [reply test/functional/feature_bind_extra.py]: Finally I figured out what is going on!

`test_ipv6_local()` tries to connect to `::1` in order to check if the system has IPv6 support.

Some (all?) Travis machines do support IPv6, but `::1` is not configured on the loopback interface (!?). On those machines connecting to `::1` fails with `[Errno 99] Cannot assign requested address`. However it is possible to successfully bind on `::`.

`bitcoind` tries to bind on `::` and `::1` and the above makes it difficult for the test to set reasonable expectations.

Resolved by ignoring all IPv6 addresses.

File: src/net.cpp
Comment (vasild) [reply src/net.cpp:2404]: Yeah. https://github.com/bitcoin/bitcoin/pull/19991#discussion_r498721063

I think it is reasonable to inform the user that his request couldn't be fulfilled. E.g. if he specifies `-bind=1.2.3.4:8334=onion` and we can't bind on that address and port.

**Comment by vasild** [comment]
Fixed the test failure on Travis (confirmed on my local fork).

File: src/net.cpp
Comment (hebasto) [reply src/net.cpp:2404]: > @hebasto WDYT, just an oversight?

Good catch! I think the `BF_REPORT_ERROR` flag was missed for no reason.

**Comment by hebasto** [comment]
>  - If only `-bind=addr` is given (without `-bind=...=onion`) then we
>   would bind to `addr` _and_ to `127.0.0.1:8334` in addition...

One could _expect_ binding to the default `127.0.0.1:8334` as no `-bind=...=onion` is provided explicitly.

> ... which may
>   be unexpected assuming the perception of `-bind=addr` is
>   "bind _only_ to `addr`".

If perception of an option is ambiguous the option docs require improvement, no?

>  - If only `-bind=addr=onion` is given (without ordinary `-bind=`)
>   then we would bind to `addr` _and_ to `0.0.0.0` in addition.

Again, binding to the default `0.0.0.0:8333` is reasonable as no `-bind=...` is provided explicitly.

I understand, that all statements above look like `-bind=...` and `-bind=...=onion` are two independent options. Are they?

**Comment by vasild** [comment]
> I understand, that all statements above look like `-bind=...` and `-bind=...=onion` are two independent options. Are they?

I think all 3 of `-bind=...`, `-bind=...=onion` and `-whitebind=` should be considered "the same" because all of them control where to listen for P2P connections.

For example, if `-whitebind=` is present and `-bind=` is not, then we do not bind on `0.0.0.0` anymore:

```
$ bitcoind -whitebind=1.1.1.1:8888
...
$ sockstat -l |grep bitcoind
vd       bitcoind   46471 6  tcp4   127.0.0.1:8332        *:*
vd       bitcoind   46471 19 tcp4   1.1.1.1:8888          *:*
```

**Comment by hebasto** [comment]
> For example, if `-whitebind=` is present and `-bind=` is not, then we do not bind on `0.0.0.0` anymore:

This behavior seems to need to be properly documented as well :)

**Comment by vasild** [comment]
This PR seems to have derailed in direction of documentation, but I think the code needs to be changed.

Take any popular server - Apache, PostgreSQL, MySQL, OpenSSH, Tor, Bitcoin Core 0.20.1 (but not the current `master`). They all listen for some service and if the user explicitly specifies to listen/bind only on a given address:port, they would not listen for that service on other addresses/ports in addition.

Apache, for example has `Listen 80` which binds on `0.0.0.0:80`, but when changed to `Listen 1.1.1.1:80` it would only bind on `1.1.1.1:80`.

The current `master` does not abide to the above and even if requested to bind on a specific address:port to listen for P2P, it may bind to `127.0.0.1` or `0.0.0.0` in addition (details are in the OP).

**Comment by MarcoFalke** [comment]
The 0.21 milestone has been removed for now. And this is not a regression anyway, AFAICT

**Comment by wodry** [comment]
> I understand, that all statements above look like `-bind=...` and `-bind=...=onion` are two independent options. Are they?

It looks to me that they are.

I find the current `-bind` option documentation with the optional `=onion` suffix rather confusing.

Could it be possible to introduce a new option `-onionbind` (analog to names rpcbind, whitebind) to set the onion bind address explicitly and prevent confusing mixing?

Both `-bind` and `-onionbind` should be able to be unset (e.g. set to 'none') to explicitly disable them. E.g. if one would only want to "onion bind" and not to bind to clearnet at all, one could set `-bind=none`, while for `-onionbind` the default would take effect.

Here a suggestion of the documentation that could be then, which I think would be clearer:
```
  -bind=<addr>[:<port>]
       Bind to given address and always listen on it (default: 0.0.0.0). Use
       [host]:port notation for IPv6.

  -onionbind=<addr>[:<port>]
       Bind onion to given address, always listen on it and tag any incoming
       connections to that address and port as incoming Tor connections.
       (defaults: 127.0.0.1:8334, testnet: 127.0.0.1:18334,
       signet: 127.0.0.1:38334, regtest: 127.0.0.1:18445)
       Use [host]:port notation for IPv6.

```
Unfortunately, I only now came to know of to the `=onion` suffix introduced in #19991.

**Comment by vasild** [comment]
Rebased due to conflicts.

@wodry what you are suggesting is out of the scope of this PR.

**Comment by vasild** [comment]
da8e632fb...9580d0605: rebase due to conflicts

**Comment by vasild** [comment]
`9580d06052...a6b8fc9012`: rebase due to conflicts

**Comment by jonatack** [comment]
Concept ACK

**Comment by laanwj** [comment]
The idea for extending the `-bind` syntax, instead of adding `-onionbind` was that not every supported overlay network needs its own set  of options. I think this was a good idea in itself.

**Review by jonatack — COMMENTED** [review]
ACK a6b8fc90129be2ae78f5d38a537dc64353124439 code review, debug built rebased to current master 3a2c84a6b5144f4ee1181, some light testing, ran the test repeatedly with various added `pprint` statements, verified that the test fails on master.

**Comment by achow101** [comment]
ACK a6b8fc90129be2ae78f5d38a537dc64353124439

Reviewed code and verified that the test fails on master and passes with this PR.

File: src/init.cpp
Comment (jonatack) [inline src/init.cpp:1743]: see discussion at https://www.erisian.com.au/bitcoin-core-dev/log-2021-06-18.html#l-200 regarding naming or a comment here

Comment (laanwj) [reply src/init.cpp:1743]: I'm no longer sure a comment or rename is necessary. I do think the interaction with `-listen` is not completely obvious when reading this code.

**Comment by laanwj** [comment]
Code review ACK a6b8fc90129be2ae78f5d38a537dc64353124439

File: src/init.cpp
Comment (hebasto) [inline src/init.cpp:1739]: Are we still able to distinguish incoming connection via Tor from one via clearnet?

Comment (vasild) [reply src/init.cpp:1739]: In this case - no (only `-bind=addr` is given, without `-bind=...=onion`).

**Comment by hebasto** [comment]
>  - If only `-bind=addr` is given (without `-bind=...=onion`) then bind to
>       `addr` (only). If we are creating tor hidden service then use `addr`
>       as target. Same behavior as before #19991.

`-bind=addr` does not mention onion network, but it _does_ change onion-related behavior, i.e., it prevents recognizing of incoming connections that come via Tor.

This new behavior is confusing (at least to me), and I don't agree with it.

**Comment by gruve-p** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/20234/commits/a6b8fc90129be2ae78f5d38a537dc64353124439

**Comment by vasild** [comment]
@hebasto, that is similar to how it operates with other networks - if the user restricts the binds to only some networks, then all other networks are ditched (affected) wrt listening/binding.

For example, assume a machine has an IPv4 address `1.2.3.4` and an IPv6 address. By default if no `-bind` is given it would bind to both addresses. If `-bind=1.2.3.4` is given it would _not_ bind on the IPv6 address. Now this statement:

"`-bind=addr` does not mention IPv6, but it does change IPv6-related behavior..."

Would this change to `-bind=` description help?
```diff
  -bind=<addr>[:<port>][=onion]
-       Bind to given address...
+       Bind only to given address...
```

**Comment by hebasto** [comment]
> For example, assume a machine has an IPv4 address `1.2.3.4` and an IPv6 address. By default if no `-bind` is given it would bind to both addresses. If `-bind=1.2.3.4` is given it would _not_ bind on the IPv6 address.

Right. But my concerns are not about binding itself, rather about behavior that users might expect -- the distinguishing incoming Tor connections.

**Comment by jonatack** [comment]
> distinguishing incoming Tor connections

Right, `CNode::m_inbound_onion and CNode::ConnectedThroughNetwork()` need to remain pertinent, as the logic is used by the inbound eviction protection and by getpeerinfo, -netinfo, and the GUI peers window.

**Comment by vasild** [comment]
I see. So it is about judging which expectation to fail:

* Fail the expectation that `-bind=addr` only binds to `addr` (leave this PR unmerged). This might have security implications: if a user specifies `-bind=1.2.3.4:8333` and expects that this is the only listen address:port and sets up his firewall rules according to that expectation. Especially that this was the case before Oct 2020 (see below).

* Fail the expectation that we distinguish incoming Tor connections if `-bind=addr` is given (merge this PR). We do that only after #19991 which was merged 9 months ago (in Oct 2020). Btw, in peer eviction we semi-distinguish Tor connections even in that case - we protect localhost peers in addition to Tor ones.

**Comment by vasild** [comment]
@hebasto, what do you think about the second change in this PR? That is - the second item from the PR description: "* If only `-bind=addr=onion` is given...". If that is not contentious then maybe it would make sense that I split the PR in two.

**Comment by hebasto** [comment]
> @hebasto, what do you think about the second change in this PR? That is - the second item from the PR description: "* If only `-bind=addr=onion` is given...". If that is not contentious then maybe it would make sense that I split the PR in two.

sgtm

**Comment by vasild** [comment]
` a6b8fc9...a004833`: remove the controversial part of this PR (the behavior when `-bind=...` is given without `-bind=...=onion` -- set it as it is in `master`), take review suggestions

File: test/functional/feature_bind_extra.py
Comment (vasild) [reply test/functional/feature_bind_extra.py:27]: Both "regtest" and "18444" are redundant with the line that follows: `REGTEST_TOR_TARGET_PORT = 18445`

File: test/functional/feature_bind_extra.py
Comment (vasild) [reply test/functional/feature_bind_extra.py:35]: Done

File: test/functional/feature_bind_extra.py
Comment (vasild) [reply test/functional/feature_bind_extra.py]: Done

File: test/functional/feature_bind_extra.py
Comment (vasild) [reply test/functional/feature_bind_extra.py]: Done

File: test/functional/feature_bind_extra.py
Comment (vasild) [reply test/functional/feature_bind_extra.py]: Done

File: test/functional/feature_bind_extra.py
Comment (vasild) [reply test/functional/feature_bind_extra.py]: You mean that `actual_without_ipv6` contains more entries on each run of the test? Then the test should fail?

File: test/functional/feature_bind_extra.py
Comment (vasild) [reply test/functional/feature_bind_extra.py:100]: The comment would be prone to becoming outdated.

File: src/init.cpp
Comment (vasild) [reply src/init.cpp]: Removed this code.

**Comment by vasild** [comment]
> > @hebasto, what do you think about the second change in this PR? That is - the second item from the PR description: "* If only `-bind=addr=onion` is given...". If that is not contentious then maybe it would make sense that I split the PR in two.
> 
> sgtm

Done!

File: src/net.cpp
Comment (vasild) [reply src/net.cpp]: [Stripping down](https://github.com/bitcoin/bitcoin/pull/20234#issuecomment-872829592) this PR I had to remove the `BF_REPORT_ERROR` flag.

If one runs two bitcoinds with explicit `-bind=...` that do not conflict one may expect that they would start successfully. The functional testing framework expects this! However they both try to bind on the same default onion target `127.0.0.1:8334`, one of them succeeds and the other fails. That failure is muted by the absence of `BF_REPORT_ERROR`. Apparently no test cares whether that bind has failed or not. To observe the problem, in `master`, add `BF_REPORT_ERROR` and run `./test/functional/test_runner.py`, observe random tests failures with errors like "unable to bind to 127.0.0.1:8334".

(the full version of this PR used to fix that and thus the presence of `BF_REPORT_ERROR` did not cause tests to brick)

**Comment by Rspigler** [comment]
Concept ACK

File: test/functional/feature_bind_extra.py
Comment (jonatack) [reply test/functional/feature_bind_extra.py]: IIRC no, the values in the entries increased. This no longer occurs with the latest push  a0048331a1d2d.

File: test/functional/feature_bind_extra.py
Comment (jonatack) [inline test/functional/feature_bind_extra.py:87]: Per offline discussion with @vasild:

```suggestion
        #        [(any_ipv4, p2p_port(2)), (loopback_ipv4, REGTEST_TOR_TARGET_PORT)]
```

and running the commented-out tests requires not having bitcoind -regtest up due to port conflict.
```
$ netstat -na | grep 18445
tcp        0      0 127.0.0.1:18445         0.0.0.0:*               LISTEN 
```

Maybe remove the commented-out tests for now.

**Review by jonatack — COMMENTED** [review]
ACK a0048331a1d2dda24cbdc7b41041ae00bf46ee7a

Verified the first test (Node0) fails on master, the second test passes (perhaps inverse their order).

**Comment by vasild** [comment]
`a0048331a1...2feec3ce31`: remove commented tests - they can be resurrected later if needed and contained a typo (thanks, @jonatack for testing even commented out stuff!)

File: test/functional/feature_bind_extra.py
Comment (vasild) [reply test/functional/feature_bind_extra.py]: Done.

**Comment by laanwj** [comment]
Code review ACK 2feec3ce3130961f98ceb030951d0e46d3b9096c

**Comment by jonatack** [comment]
utACK 2feec3ce3130961f98ceb030951d0e46d3b9096c per `git diff a004833 2feec3c`

**Comment by jonatack** [comment]
Note: This PR is tagged for 0.22, but 2 Tor/I2P PRs that should probably be priority for 0.22 but are not tagged are https://github.com/bitcoin/bitcoin/pull/22179 and https://github.com/bitcoin/bitcoin/pull/22211.

**Review by hebasto — APPROVED** [review]
ACK 2feec3ce3130961f98ceb030951d0e46d3b9096c, tested on Linux Mint 20.1 (x86_64):
```
$ src/bitcoind -signet
...
2021-07-08T14:50:35Z [init] Bound to 127.0.0.1:38334
2021-07-08T14:50:35Z [init] Bound to [::]:38333
2021-07-08T14:50:35Z [init] Bound to 0.0.0.0:38333
...
```
```
$ src/bitcoind -signet -bind=127.0.0.1:38334=onion
...
2021-07-08T14:52:02Z [init] Bound to 127.0.0.1:38334
...
```

**Pull request merged**
